src/test/java/com/microsoft/sqlserver/jdbc/SSLCertificateValidation.java

```diff
@@ -0,0 +1,135 @@
+        constructor.setAccessible(true);
+        Object hsoObject = constructor.newInstance(null, tdsc, null, serverName);
+        Method method = hsoObject.getClass().getDeclaredMethod("validateServerName", String.class);
+        method.setAccessible(true);
+        
+        // Server Name = msjdbc.database.windows.net
+        // SAN = msjdbc.database.windows.net
+        // Expected result: true
+        assertTrue((boolean) method.invoke(hsoObject, "msjdbc.database.windows.net"));
+
+        // Server Name = msjdbc.database.windows.net
+        // SAN = ms*bc.database.windows.net
+        // Expected result: true
+        assertTrue((boolean) method.invoke(hsoObject, "ms*bc.database.windows.net"));
+
+        // Server Name = msjdbc.database.windows.net
+        // SAN = *bc.database.windows.net
+        // Expected result: true
+        assertTrue((boolean) method.invoke(hsoObject, "*bc.database.windows.net"));
+
+        // Server Name = msjdbc.database.windows.net
+        // SAN = ms*.database.windows.net
+        // Expected result: true
+        assertTrue((boolean) method.invoke(hsoObject, "ms*.database.windows.net"));
+
+        // Server Name = msjdbc.database.windows.net
+        // SAN = *jd*.database.windows.net
+        // Expected result: true
+        assertTrue((boolean) method.invoke(hsoObject, "*jd*.database.windows.net"));
+
+        // Server Name = msjdbc.database.windows.net
+        // SAN = ms.*.net
+        // Expected result: false
+        assertFalse((boolean) method.invoke(hsoObject, "ms.*.net"));
+
+        // Server Name = msjdbc.database.windows.net
+        // SAN = msjdbc.asd*dsa.windows.net
+        // Expected result: false
+        assertFalse((boolean) method.invoke(hsoObject, "msjdbc.asd*dsa.windows.net"));
+
+        // Server Name = msjdbc.database.windows.net
+        // SAN = .*.windows.net
+        // Expected result: false
+        assertFalse((boolean) method.invoke(hsoObject, ".*.windows.net"));
+
+        // Server Name = msjdbc.database.windows.net
+        // SAN = msjdbc.*.windows.net
+        // Expected result: false
+        assertFalse((boolean) method.invoke(hsoObject, "msjdbc.*.windows.net"));
+
+        // Server Name = msjdbc.database.windows.net
+        // SAN = *.*.windows.net
+        // Expected result: false
+        // Note: multiple wildcards are not allowed, so this case shouldn't happen, but we still make sure to fail this.
+        assertFalse((boolean) method.invoke(hsoObject, "*.*.windows.net"));
+
+        // Server Name = msjdbc.database.windows.net
+        // SAN = *.com
+        // Expected result: false
+        // A cert with * plus a top-level domain is not allowed.
+        assertFalse((boolean) method.invoke(hsoObject, "*.com"));
+
+        // Server Name = msjdbc.database.windows.net
+        // SAN = xn--caf-dma*.com
+        // Expected result: fail
+        assertFalse((boolean) method.invoke(hsoObject, "xn--caf-dma*.com"));

```

Comment:
Shouldn't this test case be moved below now that we added tests for ""xn--ms.database.windows.net"? The cert and servers don't match here anyways, we're not sure if it's rejecting if it's a wildcard with an IDN or if it's just comparing.

Reply:
I agree that we should cover xn-- case with a wildcard that actually matches the server name (but it should fail regardless since we don't do wildcard matching against xn--). I've made this change.